Zend/zend_attributes.c

```diff
@@ -315,6 +384,17 @@ void zend_register_attribute_ce(void)
 	zend_ce_allow_dynamic_properties = register_class_AllowDynamicProperties();
 	attr = zend_internal_attribute_register(zend_ce_allow_dynamic_properties, ZEND_ATTRIBUTE_TARGET_CLASS);
 	attr->validator = validate_allow_dynamic_properties;
+
+	zend_ce_sensitive_parameter = register_class_SensitiveParameter();
+	attr = zend_internal_attribute_register(zend_ce_sensitive_parameter, ZEND_ATTRIBUTE_TARGET_PARAMETER);
+
+	/* This is not an attribute, thus the register call is missing. */
+	zend_ce_sensitive_parameter_value = register_class_SensitiveParameterValue();
+	zend_ce_sensitive_parameter_value->create_object = attributes_sensitive_parameter_value_new;
+	memcpy(&attributes_object_handlers_sensitive_parameter_value, &std_object_handlers, sizeof(zend_object_handlers));
+	attributes_object_handlers_sensitive_parameter_value.offset = XtOffsetOf(attributes_sensitive_parameter_value, std);
+	attributes_object_handlers_sensitive_parameter_value.free_obj = attributes_sensitive_parameter_value_free_obj;
```

Comment:
Nit: This works correctly, but doesn't need to be overridden with the current design and may confuse people about when to use XtOffsetOf and offset.

It calls `zend_object_std_dtor` with the exact same address casted (offset=0), and zend_object_std_dtor is the default.

It'd be needed if you weren't using readonly properties but were putting things before std instead (e.g. ext/spl/spl_fixedarray.c)

Reply:
As of PHP 8.0.0, new code should better use `offset` of `XtOffsetOf`.

Reply:
```
#ifndef XtOffsetOf
# define XtOffsetOf(s_type, field) offsetof(s_type, field)
#endif
```

I assume you mean "use `offsetof` instead of `XtOffsetOf`". Good to know that it's unconditionally defined one way

Reply:
Yeah, right, "offsetof instead of XtOffsetOf".

Reply:
Let me give the reasons behind why I did it, like I did it:

1. I've used a regular readonly property in the stub so that SensitiveParameterValue behaves like a user-land object as much as possible. It should protect against *accidental* exposure of the sensitive value, but if one _explicitly_ uses Reflection to access the internal value (for whatever reason!) then it should work as if the object was implemented in userland (e.g. in a Polyfill).

As an example, this should be possible:

```
<?php

$foo = new SensitiveParameterValue('foo');

$r = new ReflectionClass($foo);
$p = $r->getProperty('value');

var_dump($p->getValue($foo));
```

2. Protecting against the `(array)` cast is only possible for native implementations and for that I need to override `get_properties_for`. While I don't need to store any additional internal values, I've added the `attributes_sensitive_parameter_value` struct with only the `std` member for consistency with other natively implemented classes, so that the SensitiveParameterValue class is no special case. Likewise I've calculated the offsets, even if I know they are always `0`, so that adding additional struct members works properly without needing to adjust multiple locations.

If you think that I should simple use `zend_object` instead of `attributes_sensitive_parameter_value` then I can adjust the code, no problem. Just wanted to give my reasons.

Regarding `offset` vs `XtOffsetOf`: Do you have an example for that?

Reply:
@TysonAndre Okay, if there's precedent then it makes sense to simplify this. I've made the change in 2fd9a602d555473c859ce48afb95976ca0c79369.

I've also added a test for reflection in 7e94af78ad294f3e6af717cd07375c8462b97545.

If it's looking good to you I'll squash those commits into the initial commit.

Reply:
I now went ahead and rebased it, as the commit links remain valid.